File: C++/20942.cpp
#include <iostream>
#include <vector>
#include <stack>
#include <cmath>
using namespace std;

stack<int> s;
bool visited1[500111];
int visited2[500111], idx(1), tmp[50000];
vector<int> front[500111], pre[500111];

int inverse(int a) {
	return 500011 - a;
}

void dfs1(int n) {
	visited1[n] = true;
	for (auto v : front[n]) {
		if (!visited1[v])
			dfs1(v);
	}
	s.push(n);
}

void dfs2(int n) {
	visited2[n] = idx;
	for (auto v : pre[n]) {
		if (!visited2[v])
			dfs2(v);
	}
}

void make(int a, int b) {
	front[a].push_back(b);
	pre[b].push_back(a);
}

int main() {
	int N, M;
	cin >> N;
	for (int i = 0; i < N; i++)
		cin >> tmp[i];
	cin >> M;
	for (int i = 0; i < N; i++) {
		if (tmp[i]) {
			int cur = i * 5;
			for (int j = 0; j < 5; j++) {
				cur++;
				if (tmp[i] & 1 << j) {
					make(inverse(cur), cur);
					continue;
				}
				make(cur, inverse(cur));
			}
		}
	}
	for (int i = 0; i < 5 * N; i += 5) {
		int tmp3 = i + 3;
		int tmp4 = i + 4;
		int tmp5 = i + 5;
		make(inverse(tmp5), tmp4);
		make(tmp5, inverse(tmp4));
		make(tmp5, inverse(tmp3));
		make(inverse(tmp4), tmp5);
		make(tmp4, inverse(tmp5));
		make(tmp3, inverse(tmp5));
	}
	for (int nothing = 0; nothing < M; nothing++) {
		char t;
		int x, y, z;
		cin >> t >> x >> y >> z;
		x = (x * 5) - 5;
		y = (y * 5) - 5;
		if (t == '&') {
			for (int j = 0; j < 5; j++) {
				x++;
				y++;
				if (z & 1 << j) {
					make(inverse(x), x);
					make(inverse(y), y);
					continue;
				}
				make(x, inverse(y));
				make(y, inverse(x));
			}
			continue;
		}
		for (int j = 0; j < 5; j++) {
			x++;
			y++;
			if (z & 1 << j) {
				make(inverse(x), y);
				make(inverse(y), x);
				continue;
			}
			make(x, inverse(x));
			make(y, inverse(y));
		}
	}

	for (int i = inverse(5 * N); i < inverse(0); i++) {
		if (!visited1[i])
			dfs1(i);
	}
	for (int i = 1; i <= 5 * N; i++) {
		if (!visited1[i])
			dfs1(i);
	}

	while (!s.empty()) {
		int cur = s.top();
		s.pop();
		if (!visited2[cur]) {
			dfs2(cur);
			idx++;
		}
	}
	int* ans = new int[5 * N + 1];
	for (int i = 1; i <= 5 * N; i++) {
		if (visited2[i] == visited2[inverse(i)]) {
			cout << 0;
			return 0;
		}
		if (visited2[i] > visited2[inverse(i)]) {
			ans[i] = 1;
		}
		else
			ans[i] = 0;
	}
	cout << 1 << '\n';
	for (int i = 1; i < 5 * N; i += 5) {
		int res = 0;
		for (int j = 0; j < 5; j++)
			res += (1 << j) * ans[i + j];
		cout << res << ' ';
	}
}

File: Python/20942.py
import sys
sys.setrecursionlimit(10 ** 9)


def dfs1(n):
    visited[n] = True
    for v in forward[n]:
        if not visited[v]:
            dfs1(v)
    stack.append(n)


def dfs2(n):
    visited[n] = idx
    for v in backward[n]:
        if visited[v] == -1:
            dfs2(v)


N = int(sys.stdin.readline())
tmp = list(map(int, sys.stdin.readline().split()))
# print(tmp)
# seat = [0]
# for i in tmp:
#     seat += i
# print(seat)
M = int(sys.stdin.readline())
forward = [set() for _ in range(10 * N + 1)]
backward = [set() for _ in range(10 * N + 1)]
for i in range(len(tmp)):
    if tmp[i]:
        cur = i * 5
        # print(tmp[i])
        for j in range(5):
            # print(j)
            cur += 1
            if tmp[i] & 1 << j:
                forward[-cur].add(cur)
                backward[cur].add(-cur)
                continue
            forward[cur].add(-cur)
            backward[-cur].add(cur)
# print(forward)
for i in range(0, 5 * N, 5):
    tmp3 = i + 3
    tmp4 = i + 4
    tmp5 = i + 5
    forward[-tmp5].add(tmp4)
    forward[tmp5].add(-tmp4)
    forward[tmp5].add(-tmp3)
    forward[-tmp4].add(tmp5)
    forward[tmp4].add(-tmp5)
    forward[tmp3].add(-tmp5)
    backward[tmp4].add(-tmp5)
    backward[-tmp4].add(tmp5)
    backward[-tmp3].add(tmp5)
    backward[tmp5].add(-tmp4)
    backward[-tmp5].add(tmp4)
    backward[-tmp5].add(tmp3)
for _ in range(M):
    # print(forward)
    t, x, y, z = sys.stdin.readline().split()
    x = int(x) * 5 - 5
    y = int(y) * 5 - 5
    # print(x, y)
    z = int(z)
    # print(z)
    if t == '&':
        # print('&')
        for j in range(5):
            x += 1
            y += 1
            if z & 1 << j:
                forward[-x].add(x)
                forward[-y].add(y)
                backward[x].add(-x)
                backward[y].add(-y)
                continue
            forward[x].add(-y)
            forward[y].add(-x)
            backward[-y].add(x)
            backward[-x].add(y)
        continue
    for j in range(5):
        x += 1
        y += 1
        if z & 1 << j:
            forward[-x].add(y)
            forward[-y].add(x)
            backward[y].add(-x)
            backward[x].add(-y)
            continue
        forward[x].add(-x)
        forward[y].add(-y)
        backward[-x].add(x)
        backward[-y].add(y)
stack = []
visited = [False] * (10 * N + 1)
for i in range(-5 * N, 0):
    if not visited[i]:
        dfs1(i)
for i in range(1, 5 * N + 1):
    if not visited[i]:
        dfs1(i)
idx = 0
visited = [-1] * (10 * N + 1)
while stack:
    cur = stack.pop()
    if visited[cur] == -1:
        dfs2(cur)
        idx += 1
ans = [0] * (5 * N + 1)
for i in range(1, 5 * N + 1):
    if visited[i] == visited[-i]:
        print(0)
        sys.exit(0)
    if visited[i] > visited[-i]:
        ans[i] = 1
print(1)
# print(seat)
# print(visited)
for i in range(1, 5 * N, 5):
    res = 0
    for j in range(5):
        res += (1 << j) * ans[i + j]
    print(res, end=' ')
